libs/core/algorithms/include/hpx/parallel/algorithms/uninitialized_relocate.hpp

```diff
@@ -469,6 +467,64 @@ namespace hpx::parallel {
         };
         /// \endcond
 
+        /////////////////////////////////////////////////////////////////////////////
+        // uninitialized_relocate_backward
+        /// \cond NOINTERNAL
+        template <typename IterPair>
+        struct uninitialized_relocate_backward
+          : public algorithm<uninitialized_relocate_backward<IterPair>,
+                IterPair>
+        {
+            constexpr uninitialized_relocate_backward() noexcept
+              : algorithm<uninitialized_relocate_backward, IterPair>(
+                    "uninitialized_relocate_backward")
+            {
+            }
+
+            // non vectorized overload
+            template <typename ExPolicy, typename BiIter1, typename BiIter2,
+                // clang-format off
+                HPX_CONCEPT_REQUIRES_(
+                    hpx::is_sequenced_execution_policy_v<ExPolicy>&&
+                    hpx::traits::is_bidirectional_iterator_v<BiIter1>&&
+                    hpx::traits::is_bidirectional_iterator_v<BiIter2>
+                )>
+            //  clang-format on
+            static util::in_out_result<BiIter1, BiIter2> sequential(
+                ExPolicy&&, BiIter1 first, BiIter1 last,
+                BiIter2 dest_last) noexcept(hpx::experimental::util::detail::
+                relocation_traits<BiIter1, BiIter2>::is_noexcept_relocatable_v)
+            {
+                return util::in_out_result<BiIter1, BiIter2>{first,
+                    hpx::experimental::util::uninitialized_relocate_backward_primitive(
+                        first, last, dest_last)};
+            }
+
+            template <typename ExPolicy, typename BiIter1, typename BiIter2,
+            // clang-format off
+                HPX_CONCEPT_REQUIRES_(
+                    hpx::is_execution_policy_v<ExPolicy>&&
+                    hpx::traits::is_bidirectional_iterator_v<BiIter1>&&
+                    hpx::traits::is_bidirectional_iterator_v<BiIter2>
+                )>
+            // clang-format on
+            static util::detail::algorithm_result_t<ExPolicy,
+                util::in_out_result<BiIter1, BiIter2>>
+            parallel(ExPolicy&& policy, BiIter1 first, BiIter1 last,
+                BiIter2 dest_last) noexcept(hpx::experimental::util::detail::
+                    relocation_traits<BiIter1,
+                        BiIter2>::is_noexcept_relocatable_v)
+            {
+                auto count = std::distance(first, last);
+
+                auto dest_first = std::prev(dest_last, count);
+
+                return parallel_uninitialized_relocate_n(
+                    HPX_FORWARD(ExPolicy, policy), first, count, dest_first);
```

Comment:
Whether this is OK depends on the preconditions of your `parallel` algorithms. I _expect_ that this is not OK. The important thing IMO is that we want this to work:
```
Widget a[10] = {1,2,3,4,5,6,7,8,9,destroyed};
std::uninitialized_relocate_backward(a, a+9, a+10);
assert(a == {destroyed,1,2,3,4,5,6,7,8,9});
```
(modulo syntax errors and language-lawyering over the lifetimes of those Widget objects). If you copy them in the forward direction, the output will be something like
```
assert(a == {destroyed,1,1,1,1,1,1,1,1,1});
```
So the idea of having `uninitialized_relocate_backward` dispatch to `uninitialized_relocate_n` smells wrong to me.

Reply:
You are correct in pointing this out, however I should look into how (if at all) parallelizable this would be, while retaining the correct ordering.

Reply:
Naïvely, it's not parallelizable at all _unless_ you can prove (by pointer comparisons) that the two ranges don't overlap. If they don't overlap, then this implementation is fine. It occurs to me that HPX already has a `std::copy{,_backward}` implementation; you should look at what it does, and copy that strategy.
(It might be that HPX's `std::copy{,_backward}` also completely ignore the overlapping issue. If so, then I think it would be perfectly fine for you to copy that ignorance here. `uninitialized_relocate_backward` is not intended to be any _more_ difficult to implement than `copy_backward`.)

Reply:
From https://github.com/STEllAR-GROUP/hpx/pull/6364#discussion_r1413166378:
> In standard C++20 we have
>
>   -  the non-parallel std::copy_backward, which forbids overlap
>   - the ExecutionPolicy overload of std::copy_backward, which also forbids overlap
> 
> and thus {parallel_,}vector::erase fundamentally isn't allowed to use std::copy_backward.
> D1144R10 currently proposes
>
>   - a non-parallel std::uninitialized_relocate_backward which permits overlap
>   - an ExecutionPolicy overload of std::uninitialized_relocate_backward, which also permits overlap

And noting that:
>uninitialized_relocate_backward is not intended to be any more difficult to implement than copy_backward.

Are you proposing permitting overlapping ranges with parallel `copy`, or breaking the symmetry between `copy` and `relocation`?

I can experiment with implementing and benchmarking a parallel relocation algorithm for overlapping ranges.

Reply:
I'm proposing breaking the symmetry between `copy` and `uninitialized_relocate`. (But I admit that's bad; and I also admit I still don't know why `copy` forbids overlap in the first place.)

Reply:
My view is:
- `uninitialized_relocate{,_n}` should permit "shift left" overlap, as needed by `vector::erase`.
- `uninitialized_relocate_backward` should permit "shift right" overlap, as needed by `vector::insert`.

My proposed wording reflects this since [P1144R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1144r6.html#wording-uninitialized.relocate): the behavior of `uninitialized_relocate` is "as if" by a plain old for-loop relocating each element in sequence from first to last, left to right.